Detectors/Vertexing/src/SVertexer.cxx

```diff
@@ -389,25 +437,59 @@ bool SVertexer::checkV0(const TrackCand& seedP, const TrackCand& seedN, int iP,
       goodHyp = hypCheckStatus[ipid] = true;
     }
   }
-  if (!goodHyp && mSVParams->checkV0Hypothesis) {
-    LOG(debug) << "RejHypo";
-    return false;
+
+  // apply mass selections for 3-body decay
+  bool good3bodyV0Hyp = false;
+  for (int ipid = 2; ipid < 4; ipid++){
+    float massForLambdaHyp = mV0Hyps[ipid].calcMass(p2Pos, p2Neg, p2V0);
+    if ( massForLambdaHyp - mV0Hyps[ipid].getMassV0Hyp() < mV0Hyps[ipid].getMargin(ptV0) ) {
+      good3bodyV0Hyp = true;
+      break;
+    }
   }
 
+  // we want to reconstruct the 3 body decay of hypernuclei starting from the V0 of a proton and a pion (e.g. H3L->d + (p + pi-), or He4L->He3 + (p + pi-)))
+  bool checkFor3BodyDecays = mEnable3BodyDecays && (!mSVParams->checkV0Hypothesis || good3bodyV0Hyp) && (pt2V0 > 0.5);
+  bool rejectAfter3BodyCheck = false; // To reject v0s which can be 3-body decay candidates but not cascade or v0
   bool checkForCascade = mEnableCascades && r2v0 < mMaxR2ToMeanVertexCascV0 && (!mSVParams->checkV0Hypothesis || (hypCheckStatus[HypV0::Lambda] || hypCheckStatus[HypV0::AntiLambda]));
   bool rejectIfNotCascade = false;
+
+  if (!goodHyp && mSVParams->checkV0Hypothesis) {
+    LOG(debug) << "RejHypo";
+    if (!checkFor3BodyDecays){
```

Comment:
likely the checkForCascade will be false here, but better have it explicitly checked

Reply:
Added!